### PR TITLE
Add osu!track presence

### PR DESCRIPTION
**Pull request by blacbrue**
Proof that it works:

Homepage

wOkRjelzF8

User statistics

aVDPLU3GDi

User statistics (Taiko)

sOt6415SHY

User statistics (Catch The Beat)

VHXdGa1DJJ

User statistics (Mania)

qVbpVCQmBI

Best Plays of osu! by PP Earned

7H5VJwHhC1

Best Plays of osu! by PP Earned (Taiko)

65vbA6yVOT

Best Plays of osu! by PP Earned (Catch The Beat

9ps0UAmyJa

Best Plays of osu! by PP Earned (Mania)

yIE6Mjf2wu

file structure in /b/ dir (idk why did I add this)

cpAOz2xMFW

Info bout ads

lu7L3LoN9K

osu!track's discord bots

ozSrCbatBl

top 10 pp plays of osu as of 10/10/2014

mJ9BiPTgiT

osu!track's irc bot documentation (with index.php at the end of the pathname)

yZe7WpMq6i

osu!track's irc bot documentation (without index.php at the end of the pathname)

dI5eZ9yFBe

osu!track's mailerbot reference

yd9AYwaBZQ

previous one fked up

### CODE REVIEW

File: websites/O/osu!track/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
+    "$schema": "https://schemas.premid.app/metadata/1.3",
+    "service": "osu!track",
+    "author": {
+      "name": "peko",
+      "id": "398752205043400724"
+    },
+    "category": "games",
+    "logo": "https://i.imgur.com/8QvoHe1.png",
```
Comment (SlowLife1661) [inline websites/O/osu!track/dist/metadata.json:9]: Must be at least 512x512

File: websites/O/osu!track/presence.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  else if (pathnames === "/osutrack/b/pp/") {
+    presenceData.details = "Viewing at Top 10 PP Plays of osu!";
+    presenceData.state = "As of 10/10/2014";
+  } else if (pathnames === "/osutrack/updater/") 
+    presenceData.details = "Viewing at the IRC Bot's Documentation";
+  else if (pathnames === "/osutrack/updater/index.php") 
+    presenceData.details = "Viewing at the IRC Bot's Documentation";
+  
+
+	if (presenceData.details === null) {
```
Comment (SlowLife1661) [inline websites/O/osu!track/presence.ts:103]:
```suggestion
	if (!presenceData.details) {
```

**Review by Slowlife01 — COMMENTED** [review]
Would be good if you run prettier on it

File: websites/O/osu!track/dist/metadata.json (hunk 1 above)
Comment (jaydenkoh) [reply websites/O/osu!track/dist/metadata.json:9]: k

File: websites/O/osu!track/presence.ts (hunk 2 above)
Comment (jaydenkoh) [reply websites/O/osu!track/presence.ts:103]: k

**Comment by jaydenkoh** [comment]
> Would be good if you run prettier on it

`["INFO" - 9:22:40 pm] Extension Name: esbenp.prettier-vscode.
["INFO" - 9:22:40 pm] Extension Version: 8.0.1.
["INFO" - 9:22:40 pm] Attempted to load Prettier module from d:\Presences
["ERROR" - 9:22:40 pm] Failed to load module. If you have prettier or plugins referenced in package.json, ensure you have run `npm install`
["ERROR" - 9:22:40 pm] Cannot find module 'prettier' from 'd:\Presences'
Error: Cannot find module 'prettier' from 'd:\Presences'
	at Function.e.exports [as sync] (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:43599)
	at t.ModuleResolver.findPkg (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:5790)
	at t.ModuleResolver.<anonymous> (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:3671)
	at Generator.next (<anonymous>)
	at c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:2894
	at new Promise (<anonymous>)
	at n (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:2639)
	at t.ModuleResolver.getPrettierInstance (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:3403)
	at t.default.<anonymous> (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:8998)
	at Generator.next (<anonymous>)
	at c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:7201
	at new Promise (<anonymous>)
	at n (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:6946)
	at t.default.handleActiveTextEditorChanged (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:8664)
	at t.default.registerDisposables (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:11734)
	at e.activate (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:73978)
	at Function._callActivateOptional (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:14966)
	at Function._callActivate (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:14639)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:89:12799
	at runMicrotasks (<anonymous>)
	at processTicksAndRejections (internal/process/task_queues.js:93:5)
	at async Promise.all (index 0)
["INFO" - 9:24:45 pm] Attempted to load Prettier module from d:\Presences
["ERROR" - 9:24:45 pm] Failed to load module. If you have prettier or plugins referenced in package.json, ensure you have run `npm install`
["ERROR" - 9:24:45 pm] Cannot find module 'prettier' from 'd:\Presences'
Error: Cannot find module 'prettier' from 'd:\Presences'
	at Function.e.exports [as sync] (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:43599)
	at t.ModuleResolver.findPkg (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:5790)
	at t.ModuleResolver.<anonymous> (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:3671)
	at Generator.next (<anonymous>)
	at c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:2894
	at new Promise (<anonymous>)
	at n (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:2639)
	at t.ModuleResolver.getPrettierInstance (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:3403)
	at t.default.<anonymous> (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:8998)
	at Generator.next (<anonymous>)
	at c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:7201
	at new Promise (<anonymous>)
	at n (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:6946)
	at handleActiveTextEditorChanged (c:\Users\lione\.vscode\extensions\esbenp.prettier-vscode-8.0.1\dist\extension.js:1:8664)
	at l.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:57:1835)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:85:46664
	at l.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:57:1835)
	at E.acceptDocumentsAndEditorsDelta (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:84:148554)
	at E.$acceptDocumentsAndEditorsDelta (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:84:146684)
	at d._doInvokeHandler (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:88:12815)
	at d._invokeHandler (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:88:12507)
	at d._receiveRequest (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:88:11160)
	at d._receiveOneMessage (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:88:9960)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:88:8064
	at l.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:57:1835)
	at E.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:15634)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:103:29816
	at l.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:57:1835)
	at E.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:15634)
	at t._receiveMessage (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:20892)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:17778
	at l.fire (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:57:1835)
	at c.acceptChunk (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:12995)
	at c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:65:12343
	at Socket.b (c:\Users\lione\AppData\Local\Programs\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:103:13287)
	at Socket.emit (events.js:315:20)
	at addChunk (internal/streams/readable.js:309:12)
	at readableAddChunk (internal/streams/readable.js:284:9)
	at Socket.Readable.push (internal/streams/readable.js:223:10)
	at Pipe.onStreamRead (internal/stream_base_commons.js:188:23)
`

**Comment by ririxidev** [comment]
Do you even know how prettier works? Remove this comment and better install dependencies on project and then try bruh

**Review by rxri — CHANGES_REQUESTED** [review]
can i ask why you have duplicated else-if and you can do it in better unique way

**Comment by jaydenkoh** [comment]
> can i ask why you have duplicated else-if and you can do it in better unique way

if I use switch case break the presence wouldn't show up

**Comment by ririxidev** [comment]
ugh it's still unecessary duplicated cases, just use `.includes` function and u r happy

**Comment by jaydenkoh** [comment]
ok

**Comment by jaydenkoh** [comment]
> ugh it's still unecessary duplicated cases, just use `.includes` function and u r happy

but if I use includes on pathnames.includes(`/osutrack/user/${urlSplit[5]}` then the pathnames.includes(`/osutrack/user/${urlSplit[5]}/taiko`, pathnames.includes(`/osutrack/user/${urlSplit[5]}/ctb` and pathnames.includes(`/osutrack/user/${urlSplit[5]}/mania` presence will just stay the same

**Comment by jaydenkoh** [comment]
nvm i just used `||`

File: websites/O/osu!track/dist/metadata.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      "osutrack",
+      "osu-stats",
+      "osu-top-plays",
+      "osu-taiko",
+      "osu-mania",
+      "osu-catch-the-beat"
+    ],
+    "url": "ameobea.me",
+    "description": {
+      "en": "Site where players of the game osu! can track their progress and view the progression of others. Pulls data from the osu! API, storing time series data and creating a variety of charts, plots, statistics, and other visualizations for a variety of different metrics. An in-game chat bot and Discord bot were also written to allow users to update their stats from in-game. Thousands of players a month continue to use osu!track."
```
Comment (Bas950) [inline websites/O/osu!track/dist/metadata.json:24]: get a shorter description

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/O/osu!track/dist/metadata.json (hunk 3 above)
Comment (jaydenkoh) [reply websites/O/osu!track/dist/metadata.json:24]: k

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/O/osu!track/presence.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    presenceData.details = "Viewing at the IRC Bot's Documentation";
+
+  if (!presenceData.details) {
+    //This will fire if you do not set presence details
+    presence.setTrayTitle(); //Clears the tray title for mac users
+    presence.setActivity(); /*Update the presence with no data, therefore clearing it and making the large image the Discord Application icon, and the text the Discord Application name*/
+  } else {
+    //This will fire if you set presence details
+    presence.setActivity(presenceData); //Update the presence with all the values from the presenceData object
+  }
```
[the remark this answers is not in the document]
Comment (jaydenkoh) [reply websites/O/osu!track/presence.ts:88]: k

File: websites/O/osu!track/presence.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  }),
+  timestampe = Math.floor(Date.now() / 1000);
+
+presence.on("UpdateData", async () => {
+  const presenceData: PresenceData = {
+    largeImageKey: "logo",
+    startTimestamp: timestampe
+  },
+    pathnames = location.pathname,
+    urlSplit = document.URL.split("/"),
```
Comment (Bas950) [inline websites/O/osu!track/presence.ts:12]: Seems like you always use [5], maybe use array deconstructing?

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/O/osu!track/presence.ts (hunk 5 above)
Comment (jaydenkoh) [reply websites/O/osu!track/presence.ts:12]: i forgot how to do it but ill try

Comment (Bas950) [reply websites/O/osu!track/presence.ts:12]: [, , , , urlSplit] = document.URL.split("/")

Comment (jaydenkoh) [reply websites/O/osu!track/presence.ts:12]: figured it out

Comment (jaydenkoh) [reply websites/O/osu!track/presence.ts]: i did 
`urlSplitRaw = document.URL.split("/"),
[, , , , , urlSplit] = urlSplitRaw,`

worked for me

File: websites/O/osu!track/presence.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  timestampe = Math.floor(Date.now() / 1000);
+
+presence.on("UpdateData", async () => {
+  const presenceData: PresenceData = {
+    largeImageKey: "logo",
+    startTimestamp: timestampe
+  },
+    pathnames = location.pathname,
+    urlSplitRaw = document.URL.split("/"),
+    [, , , , , urlSplit] = urlSplitRaw,
```
Comment (Bas950) [inline websites/O/osu!track/presence.ts:13]:
```suggestion
    [, , , , , urlSplit] = document.URL.split("/"),
```

**Review by Bas950 — COMMENTED** [review]
That should work too

**Review by Slowlife01 — APPROVED** [review]
Looks good to me

**Review by Bas950 — APPROVED** [review]

**Pull request merged**